internal/storage/image.go

```diff
@@ -398,6 +407,75 @@ func (svc *imageService) ImageStatus(systemContext *types.SystemContext, nameOrI
 	return &result, nil
 }
 
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string) error {
+	log.Debugf(ctx, "Getting image status of config image: %s", configImage)
+	configImageStatus, err := svc.ImageStatus(systemContext, configImage)
+	if err != nil {
+		return fmt.Errorf("image status from config image: %w", err)
+	}
+
+	log.Debugf(ctx, "Getting image status of user specified image: %s", configImage)
+	userSpecifiedImageStatus, err := svc.ImageStatus(systemContext, userSpecifiedImage)
+	if err != nil {
+		return fmt.Errorf("image status from user specified image: %w", err)
+	}
+
+	log.Debugf(ctx, "Comparing digest %s with %s, both must be equal.", configImageStatus.Digest, userSpecifiedImageStatus.Digest)
+	if configImageStatus.Digest != userSpecifiedImageStatus.Digest {
+		return fmt.Errorf(
+			"config image digest does not match user specified image: %s != %s",
+			configImageStatus.Digest, userSpecifiedImageStatus.Digest,
+		)
+	}
+
+	// Check the cache first
+	var policyContext *signature.PolicyContext
+	item := svc.policyContexts.Get(systemContext.SignaturePolicyPath)
+	if item != nil {
+		log.Debugf(ctx, "Using policy context from cache")
+		policyContext = item.Value()
+	} else {
+		log.Debugf(ctx, "Creating policy context")
+		policy, err := signature.DefaultPolicy(systemContext)
+		if err != nil {
+			return fmt.Errorf("get default policy: %w", err)
+		}
+
+		policyContext, err = signature.NewPolicyContext(policy)
+		if err != nil {
+			return fmt.Errorf("create policy context: %w", err)
+		}
+
+		log.Debugf(ctx, "Caching policy context")
+		svc.policyContexts.Set(systemContext.SignaturePolicyPath, policyContext, ttlcache.DefaultTTL)
+	}
+
+	log.Debugf(ctx, "Parsing docker reference from %s", userSpecifiedImage)
+	imageRef, err := docker.ParseReference("//" + userSpecifiedImage)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf(ctx, "Creating image source")
+	imageSrc, err := imageRef.NewImageSource(ctx, systemContext)
+	if err != nil {
+		return fmt.Errorf("create image source: %w", err)
+	}
+
+	log.Debugf(ctx, "Check if image is allowed to run")
+	unparsedImage := cimage.UnparsedInstance(imageSrc, &userSpecifiedImageStatus.Digest)
+	if _, err := policyContext.IsRunningImageAllowed(ctx, unparsedImage); err != nil {
```

Comment:
⚠️ 

This checks whether the on-registry image is correctly signed. But _nothing_ connects the on-registry image artifact with the already-locally-pulled `config.Image`, i.e. this does not verify anything about the image that is about to be run.